src/FrameReflower/Block.php

```diff
@@ -792,26 +792,32 @@ function reflow(BlockFrameDecorator $block = null)
 
         // Set the containing blocks and reflow each child
         foreach ($this->_frame->get_children() as $child) {
-
-            // Bail out if the page is full
-            if ($page->is_full()) {
-                break;
-            }
-
             $child->set_containing_block($cb_x, $cb_y, $w, $cb_h);
 
             $this->process_clear($child);
 
             $child->reflow($this->_frame);
 
-            // Don't add the child to the line if a page break has occurred
-            if ($page->check_page_break($child)) {
-                break;
-            }
-
             $this->process_float($child, $cb_x, $w);
         }
 
+        // Go back through and handle any page breaks
+        foreach ($this->_frame->get_children() as $child) {
+            // Skip children that are empty text nodes
+            if (!$child->is_text_node() ||
+                $child->get_node()->nodeValue != ""
+            ) {
+                if ($page->check_page_break($child)) {
+                    break;
+                }
+
+                // Bail out if the page is full
+                if ($page->is_full()) {
+                    break;
+                }
+            }
+        }
+
```

Comment:
This change is problematic with the current infrastructure: A page break in a child element (when reflowing the child) can now result in siblings before it staying on the original page, even though there is not enough space for them; they are checked after the page break has already occurred and so no page break is possible anymore.

Reply:
To illustrate that see the following HTML:

```html
<!DOCTYPE html>
<html>

<head>
<meta charset="UTF-8">
<style type="text/css">
@page {
	size: 400pt 300pt;
	margin: 50pt;
}

body {
	margin: 0;
}

.block {
	height: 200pt;
	background-color: red;
}
</style>
</head>

<body><div class="block"></div><div class="block"></div><div class="block"></div><div class="block"></div><div class="block"><div></div></div></body>

</html>
```

An the before and after rendering:
[swallow-before.pdf](https://github.com/dompdf/dompdf/files/6919548/swallow-before.pdf)
[swallow-after.pdf](https://github.com/dompdf/dompdf/files/6919550/swallow-after.pdf)

Reply:
Maybe not quite right but I think something along these lines is what we'll need to do in order to support widows. Otherwise we don't have a valid count of the number of lines.

Reply:
This also fails if a page break is forced using page-break styling.

For sure to fully support this would require a deep rework of how the document is reflowed, e.g. reflow the content then page it instead of doing both at the same time, maybe followed by additional steps. Changing the process this way might allow us to implement some functionality that we might have trouble with otherwise. This would be a major re-work, though, so best to pursue in a future release.